website/guides/border.md

```diff
@@ -0,0 +1,164 @@
+
+this adds up to 16 bytes, but if a packet doesnt use all 16 bytes, the unused bytes can be set to any value
+
+you can use [existing code](https://github.com/zlago/violence-gbc/blob/11cfdb6ee8a35e042fa9712484d814e0961cea7c/src/sub.sm83#L413-L463) or write your own code for that
+
+[documentation](https://gbdev.io/pandocs/SGB_Command_Packet.html) if youre doing the latter
+
+**you should wait 4 frames between each packet**
+
+## TRN
+
+some packets will tell the SGB to copy the contents of the screen somewhere, including the packets needed for SGB borders
+
+for that to function properly, you must:
+1. set `BGP` to `$e4` and `LCDC` to `$91` (screen enabled, BG uses tiles `$8000`-`$8fff` and tilemap `$9800`, WIN and OBJ disabled, BG enabled)
+2. set `SCX` and `SCY` to `$00`
+3. the tilemap consists of `$00`, `$01`..`$13`, 12 bytes padding (offscreen), `$14`..`$27`, padding, repeat untill `$ff` (inclusive)
+4. data at `$8000`-`$8fff`
+
+you can do 1, 2 and 3 via [this snippet](https://github.com/zlago/snek-gbc/blob/baef0369f57d2b0d58316cb1c28c6cc22475a6c9/code/init.sm83#L208-L230)
+
+**you must first upload the data then send the TRN packet, you must wait ~8 frames after each TRN** (doesnt "stack" with the 4 frame packet delay)
+
+## detection
+
+[snippet](https://github.com/zlago/snek-gbc/blob/baef0369f57d2b0d58316cb1c28c6cc22475a6c9/code/init.sm83#L167-L196), change `.init` to wherever code should jump on *not* sgb, and the code following the snippet is what should get run if this is an SGB
+
+now let me explain what the snippet actually does (`/**/` comments)
+```sm83asm
+; test for SGB
+xor a /* first it enables STAT interrupt, since snek-gbc happens to just have a `reti` as the handler */
+ldh [rLYC], a
+ld a, STATF_LYC
+ldh [rSTAT], a
+ld a, IEF_STAT
+ldh [rIE], a
+ei
+	; wait for SGB
+	ld b, 12 /* you must wait 12 or more frames before trying to send a packet */
+	:halt
+	dec b
+	jr nz, :-
+	; enable multi
+	ld hl, Packets.mlt /* send a `MLT_REQ` */
+	call Packet
+	rept 4 /* then wait 4 more frames, since the SNES wont "read" the packet instantly */
+		halt
+	endr
+	; check if SGB responds
+		/* and now we actually try to detect the SGB
+		setting `P1.5` to low then high advances the "read player"
+		setting `P1.4` and `P1.5` high will make the SGB return which player is currently selected in `P1.0` and `P1.1` */
+		lb bc, 5, LOW(rP1) /* b is loaded with 5 which is how many times we try to check if the player changes to anything but P1 */
+		:ld a, P1F_4 /* try to advance player */
+		ldh [c], a
+		ld a, P1F_4|P1F_5 /* then set P1 to return the current player */
+		ldh [c], a
+		ldh a, [c]
+		dec b
+		jp z, .init ; give up /* if 5 (4 actually) attempts pass, assume this isnt an SGB */
+		cp $ff /* `P1.6` and `P1.7` always return %1, we set `P1.4` and `P1.5` to %1, and DMG, or SGB player 1, return %1111 in `P1.0`-`P1.3` */
+		jr z, :- ; try again if failed /**/
+```
```

Comment:
This is perfectly fine code for using in a game, but I don't think it's good tutorial code because it drags in a bunch of implicit assumptions that someone who is learning may not be familiar with. It's using `halt` to wait for interrupts but this relies on the specific setup listed here:

- Interrupts must be enabled during this section, whereas interrupts are often disabled during setup code.
- LCDC must be on during this code in order to generate the interrupts, whereas LCDC is often disabled during initialization to make VRAM accessible.
- It relies in the specific setup of `STAT` which is given here as "magic values".
- It relies on the STAT interrupt being a single `reti` or similar.

This (imo) makes this code fragile for use by a beginner unless they understand how all the pieces fit together. (By fragile I mean that a change somewhere else can easily cause this code to stop working.) It would be easy for me, who has been doing this for 20 years at this point, to say that this is good code because I can understand, see that it works, or otherwise know exactly what to do if it doesn't work. But I'm not the target audience of this code.

I would much prefer the more "traditional" approach of a busy wait loop for waiting frames, which, while not as fancy, solves all of the issues listed above.

Reply:
- doesnt the doc specify that LCD must be enabled during TRNs?
- arent waitloops that a beginner would use ineffective if LCD is off?
- are you not going to comment on the doc assuming you have means of VRAM-safe VRAM access?

Reply:
> doesnt the doc specify that LCD must be enabled during TRNs?

Sure, but this is something they must keep in mind if they might later rearrange the code to put the detection in a section with LCDC disabled. And that doesn't address any of the other issues. For example, I'm assuming you chose to use STAT interrupt instead of the more obvious VBlank interrupt in order to avoid extra complexity in the VBlank interrupt handler. But now the STAT interrupt is "used up" instead. Using a waitloop makes this code more self contained without forcing assumptions on other parts of the code.

> arent waitloops that a beginner would use ineffective if LCD is off?

I'm talking about a waitloop with a counter, not a `LY` waitloop. Something like Nintendo's suggestion:
```sm83asm
    ld de,$1b58
:   nop
    nop
    nop
    dec de
    ld a,d
    or e
    jr nz,:-
    ret
```

That particular code has some awkward `nop`s in there, seemingly so that the loop takes exactly 10 M cycles and spins 7000 times for a total of 70000 M cycles for easy math. But that's the general idea.

> are you not going to comment on the doc assuming you have means of VRAM-safe VRAM access?

Safe VRAM access in itself is one of the most central concepts in GB programming, so someone who is new will be presumed to learn it sooner rather than later.

Reply:
actually, i already happened to have STAT used up for doing nothing in snek-gbc (which is where all the snippets are from), in every other project i use vblank since there the handler only does what its told to (snek-gbc was programmed with the assumption that i would never have to build upon it, dont ask me why i felt the *need* to include a dmg-only startup animation)

Reply:
So to summarize, are these the requested changes?
- Do the delay with `di` and a busy wait loop instead of vblank/STAT interrupt
- Emphasize that the display must be on, or the SGB will never receive the data

Reply:
yep, @pinobatch , sounds like it ;)